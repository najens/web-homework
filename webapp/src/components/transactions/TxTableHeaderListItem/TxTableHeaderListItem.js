import React from 'react'
import { css } from '@emotion/core'
import { ArrowDropUp, ArrowDropDown } from '@material-ui/icons'
import { string, func } from 'prop-types'

export default function TxTableHeaderListItem ({ asc, desc, heading, orderBy, refetchTransactions, setOrderBy, width }) {
  const isAsc = orderBy === asc
  const isDesc = orderBy === desc
  const isFiller = heading === ''

  function handleHeaderClick () {
    const newOrderBy = isDesc ? asc : desc
    refetchTransactions({ orderBy: newOrderBy })
    setOrderBy(newOrderBy)
  }

  return (
    <th css={headerStyle} style={{ width }}>
      {!isFiller && (
        <button css={buttonStyle} onClick={handleHeaderClick}>
          <span>{heading}</span>
          {isAsc && <ArrowDropUp style={{ fontSize: 24 }} />}
          {isDesc && <ArrowDropDown style={{ fontSize: 24 }} />}
        </button>
      )}
    </th>
  )
}

TxTableHeaderListItem.propTypes = {
  asc: string,
  desc: string,
  heading: string,
  orderBy: string,
  refetchTransactions: func,
  setOrderBy: func,
  width: string
}

const headerStyle = css`
  border: 1px solid #b4c1ec;
`

const buttonStyle = css`
  align-items: center;
  background: none;
  border: none;
  display: flex;
  flex-direction: row;
  height: 24px;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;

  &:focus {
    outline: none;
  }

  &:hover {
    cursor: pointer;
  }
`
