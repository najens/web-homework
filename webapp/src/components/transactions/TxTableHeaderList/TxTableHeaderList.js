import React, { useState } from 'react'
import TxTableHeaderListItem from '../TxTableHeaderListItem'
import { func } from 'prop-types'

export default function TxTableHeaderList ({ refetchTransactions }) {
  const [orderBy, setOrderBy] = useState('INSERTED_AT_DESC')

  const tableHeadings = [
    {
      asc: 'INSERTED_AT_ASC',
      desc: 'INSERTED_AT_DESC',
      heading: 'Date',
      key: 'date',
      width: '10%'
    },
    {
      asc: 'USER_NAME_ASC',
      desc: 'USER_NAME_DESC',
      heading: 'User',
      key: 'user',
      width: '15%'
    },
    {
      asc: 'DESCRIPTION_ASC',
      desc: 'DESCRIPTION_DESC',
      heading: 'Description',
      key: 'description',
      width: '25%'
    },
    {
      asc: 'MERCHANT_ASC',
      desc: 'MERCHANT_DESC',
      heading: 'Merchant',
      key: 'merchant',
      width: '15%'
    },
    {
      asc: 'CATEGORY_ASC',
      desc: 'CATEGORY_DESC',
      heading: 'Category',
      key: 'category',
      width: '15%'
    },
    {
      asc: '',
      desc: '',
      heading: '',
      key: 'filler',
      width: '8%'
    },
    {
      asc: 'AMOUNT_ASC',
      desc: 'AMOUNT_DESC',
      heading: 'Amount',
      key: 'amount',
      width: '12%'
    }
  ]
  return tableHeadings.map(th => {
    const {
      asc,
      desc,
      heading,
      key,
      width
    } = th

    return (
      <TxTableHeaderListItem
        asc={asc}
        desc={desc}
        heading={heading}
        key={`tx-header-${key}`}
        orderBy={orderBy}
        refetchTransactions={refetchTransactions}
        setOrderBy={setOrderBy}
        width={width}
      />
    )
  })
}

TxTableHeaderList.propTypes = {
  refetchTransactions: func
}
