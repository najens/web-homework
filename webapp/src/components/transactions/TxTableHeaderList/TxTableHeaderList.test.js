import React from 'react'
import { create, act } from 'react-test-renderer'
import TxTableHeaderListItem from '../TxTableHeaderListItem'
import TxTableHeaderList from './TxTableHeaderList'
import { ArrowDropUp, ArrowDropDown } from '@material-ui/icons'

describe('the TxTableHeaderList component', () => {
  let testRenderer

  beforeEach(() => {
    testRenderer = create(
      <TxTableHeaderList refetchTransactions={() => null} />
    )
  })

  test('it renders correctly', () => {
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  test('it renders 7 TxTableHeaderListItem children', () => {
    const testInstance = testRenderer.root
    const headers = testInstance.findAllByType(TxTableHeaderListItem)
    expect(headers).toHaveLength(7)
  })

  describe('the TxTableHeaderListItem child components', () => {
    describe('rendering arrow icons on table headers', () => {
      let firstHeader
      let secondHeader

      beforeEach(() => {
        const testRenderer = create(
          <TxTableHeaderList refetchTransactions={() => null} />
        )
        const testInstance = testRenderer.root
        const headers = testInstance.findAllByType(TxTableHeaderListItem)
        firstHeader = headers[0]
        secondHeader = headers[1]
      })

      test('the first header should render ArrowDropDown', () => {
        expect(firstHeader.findAllByType(ArrowDropDown)).toHaveLength(1)
      })

      test('the first header should not render ArrowDropUp', () => {
        expect(firstHeader.findAllByType(ArrowDropUp)).toHaveLength(0)
      })

      test('the second header should not render ArrowDropDown', () => {
        expect(secondHeader.findAllByType(ArrowDropDown)).toHaveLength(0)
      })

      test('the second header should not render ArrowDropUp', () => {
        expect(secondHeader.findAllByType(ArrowDropUp)).toHaveLength(0)
      })

      describe('first table header button click', () => {
        let firstHeader
        let secondHeader
        beforeEach(() => {
          const testRenderer = create(
            <TxTableHeaderList refetchTransactions={() => null} />
          )
          const testInstance = testRenderer.root
          const headers = testInstance.findAllByType(TxTableHeaderListItem)
          firstHeader = headers[0]
          secondHeader = headers[1]
          const button = firstHeader.findByType('button')
          act(() => button.props.onClick())
        })

        test('it should not render ArrowDropDown', () => {
          expect(firstHeader.findAllByType(ArrowDropDown)).toHaveLength(0)
        })

        test('it should render ArrowDropUp', () => {
          expect(firstHeader.findAllByType(ArrowDropUp)).toHaveLength(1)
        })

        test('the second header should not render ArrowDropDown', () => {
          expect(secondHeader.findAllByType(ArrowDropDown)).toHaveLength(0)
        })

        test('the second header should not render ArrowDropUp', () => {
          expect(secondHeader.findAllByType(ArrowDropUp)).toHaveLength(0)
        })
      })

      describe('second table header button click', () => {
        let firstHeader
        let secondHeader

        beforeEach(() => {
          const testRenderer = create(
            <TxTableHeaderList refetchTransactions={() => null} />
          )
          const testInstance = testRenderer.root
          const headers = testInstance.findAllByType(TxTableHeaderListItem)
          firstHeader = headers[0]
          secondHeader = headers[1]
          const button = secondHeader.findByType('button')
          act(() => button.props.onClick())
        })

        test('the first header should render ArrowDropDown', () => {
          expect(firstHeader.findAllByType(ArrowDropDown)).toHaveLength(0)
        })

        test('the first header should not render ArrowDropUp', () => {
          expect(firstHeader.findAllByType(ArrowDropUp)).toHaveLength(0)
        })

        test('the second header should not render ArrowDropDown', () => {
          expect(secondHeader.findAllByType(ArrowDropDown)).toHaveLength(1)
        })

        test('the second header should not render ArrowDropUp', () => {
          expect(secondHeader.findAllByType(ArrowDropUp)).toHaveLength(0)
        })
      })
    })
  })
})
