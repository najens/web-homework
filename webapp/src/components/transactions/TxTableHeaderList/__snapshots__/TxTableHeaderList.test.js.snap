// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the TxTableHeaderList component it renders correctly 1`] = `
Array [
  .emotion-1 {
  border: 1px solid #b4c1ec;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 24px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover {
  cursor: pointer;
}

<th
    className="emotion-1"
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
    >
      <span>
        Date
      </span>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        style={
          Object {
            "fontSize": 24,
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </button>
  </th>,
  .emotion-1 {
  border: 1px solid #b4c1ec;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 24px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover {
  cursor: pointer;
}

<th
    className="emotion-1"
    style={
      Object {
        "width": "15%",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
    >
      <span>
        User
      </span>
    </button>
  </th>,
  .emotion-1 {
  border: 1px solid #b4c1ec;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 24px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover {
  cursor: pointer;
}

<th
    className="emotion-1"
    style={
      Object {
        "width": "25%",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
    >
      <span>
        Description
      </span>
    </button>
  </th>,
  .emotion-1 {
  border: 1px solid #b4c1ec;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 24px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover {
  cursor: pointer;
}

<th
    className="emotion-1"
    style={
      Object {
        "width": "15%",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
    >
      <span>
        Merchant
      </span>
    </button>
  </th>,
  .emotion-1 {
  border: 1px solid #b4c1ec;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 24px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover {
  cursor: pointer;
}

<th
    className="emotion-1"
    style={
      Object {
        "width": "15%",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
    >
      <span>
        Category
      </span>
    </button>
  </th>,
  .emotion-0 {
  border: 1px solid #b4c1ec;
}

<th
    className="emotion-0"
    style={
      Object {
        "width": "8%",
      }
    }
  />,
  .emotion-1 {
  border: 1px solid #b4c1ec;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 24px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0px 6px;
  width: 100%;
  -moz-appearance: none;
  -webkit-appearance: none;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:hover {
  cursor: pointer;
}

<th
    className="emotion-1"
    style={
      Object {
        "width": "12%",
      }
    }
  >
    <button
      className="emotion-0"
      onClick={[Function]}
    >
      <span>
        Amount
      </span>
    </button>
  </th>,
]
`;
